eval/NEESEvaluator.cpp

```diff
@@ -0,0 +1,207 @@
+}
+
+void NEESEvaluator::printNeesStatistics(const std::vector<double>& nees_results, 
+                                       double preint_time) {
+    if (nees_results.empty()) {
+        std::cerr << "No valid NEES results for preintegration time " << preint_time << std::endl;
+        return;
+    }
+
+    Eigen::Map<const Eigen::VectorXd> nees_vector(nees_results.data(), nees_results.size());
+    double mean = nees_vector.mean();
+    auto sorted_results = nees_results;
+    std::sort(sorted_results.begin(), sorted_results.end());
+    double median = sorted_results[sorted_results.size()/2];
+    double variance = (nees_vector.array() - mean).square().mean();
+
+    std::cout << "ANEES (15-DOF):    mean | median | variance\n"
+              << "--------------------------------------------\n"
+              << "Results:   " << mean << " | " << median << " | " << variance << "\n";
+}
+
+NEESEvaluator::GTData NEESEvaluator::loadEurocData(const std::string& filename) {
+    std::ifstream file(filename);
+    if (!file.is_open()) {
+        throw std::runtime_error("Could not open file: " + filename);
+    }
+
+    GTData data;
+    std::string line;
+    if (!std::getline(file, line)) {
+        throw std::runtime_error("Empty file: " + filename);
+    }
+
+    double t0 = 0;
+    bool first = true;
+
+    while (std::getline(file, line)) {
+        std::vector<double> row;
+        std::stringstream ss(line);
+        std::string value;
+        while (std::getline(ss, value, ',')) {
+            row.push_back(std::stod(value));
+        }
+
+        if (first) {
+            t0 = row[0];
+            first = false;
+        }
+
+        data.timestamps.push_back(row[0] - t0);
+        
+        // Parse quaternion (w,x,y,z)
+        Rot3 rot = Rot3::Quaternion(row[1], row[2], row[3], row[4]);
+        Point3 vel(row[5], row[6], row[7]);
+        Point3 pos(row[8], row[9], row[10]);
+        data.nav_states.push_back(NavState(rot, pos, vel));
+
+        Vector3 gyro_bias(row[11], row[12], row[13]);
+        Vector3 acc_bias(row[14], row[15], row[16]);
+        data.biases.push_back(imuBias::ConstantBias(acc_bias, gyro_bias));
+
+        data.measured_omegas.push_back(Vector3(row[17], row[18], row[19]));
+        data.measured_accs.push_back(Vector3(row[20], row[21], row[22]));
+    }
+    return data;
+}
+
+void NEESEvaluator::validateData(const GTData& data) {
+    if (data.timestamps.size() < 2) {
+        throw std::runtime_error("Insufficient data points");
+    }
+}
+
+void NEESEvaluator::processAllTimeWindows(const GTData& data, 
+                                        const std::vector<double>& preint_times,
+                                        double alpha) {
+    auto params = configureImuParams(alpha);
+    double dt = data.timestamps[1] - data.timestamps[0];
+    
+    for (double preint_time : preint_times) {
+        processTimeWindow(data, params, preint_time, dt);
```

Comment:
You should choose one convention and stick to it:
- call “up” 
- ‘call “down”
you’re mixing the two. I typically prefer calling smaller functions that are defined earlier, I.e., “up”.